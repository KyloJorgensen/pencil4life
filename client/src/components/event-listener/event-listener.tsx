'use strict';

import * as React from 'react';
import './event-listener.less';

export type IEventType = 'click' | 'scroll' | 'resize';

export type IEventHandler = (event: UIEvent, ref: React.RefObject<HTMLDivElement>) => void;

export interface IEventListenerContext {
  addEventListener: (eventType: IEventType, eventHandler: IEventHandler) => boolean;
  removeEventListener: (eventType: IEventType, eventHandler: IEventHandler) => boolean;
}

const Context: IEventListenerContext =  {
	addEventListener: (eventType, eventHandler) => {
    console.error('Error: Cannot addEventListener no EventListenerContext.Provider element in parents');
    return false;
  },
	removeEventListener: (eventType, eventHandler) => {
	  console.error('Error: Cannot removeEventListener no EventListenerContext.Provider element in parents');
    return false;  
  },
}

export const EventListenerContext = React.createContext(Context);

export interface IEventListener {
  eventType: IEventType,
  eventHandler: IEventHandler;
}

export interface IEventListenerWrapperState {
  vaildEventTypes: IEventType[];
  eventListeners: IEventListener[];
}

export interface IEventListenerWrapperMethods {
  handleClick: (event: UIEvent) => void;
  handleScroll: (event: UIEvent) => void;
  handleResize: (event: UIEvent) => void;
  addEventListener: (eventType: IEventType, eventHandler: IEventHandler) => boolean;
  removeEventListener: (eventType: IEventType, eventHandler: IEventHandler) => boolean;
}

export class EventListenerWrapper extends React.Component<null, IEventListenerWrapperState> implements IEventListenerWrapperMethods {
  constructor(props) {
    super(props);
    this.state = {
      vaildEventTypes: [
        'click',
        'scroll',
        'resize'
      ],
      eventListeners: [],
    };
    this.handleClick = this.handleClick.bind(this);
    this.handleScroll = this.handleScroll.bind(this);
    this.handleResize = this.handleResize.bind(this);
    this.addEventListener = this.addEventListener.bind(this);
    this.removeEventListener = this.removeEventListener.bind(this);
  }

  eventRef: React.RefObject<HTMLDivElement> = React.createRef();

  handleClick(event) {
    this.state.eventListeners.filter(listener => listener.eventType == 'click').forEach(listener => listener.eventHandler(event, this.eventRef));
  };

  handleScroll(event) {
    this.state.eventListeners.filter(listener => listener.eventType == 'scroll').forEach(listener => listener.eventHandler(event, this.eventRef));
  }

  handleResize(event) {
    this.state.eventListeners.filter(listener => listener.eventType == 'resize').forEach(listener => listener.eventHandler(event, this.eventRef));
  }

  componentDidMount() {
    window.addEventListener('resize', this.handleResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  addEventListener(eventType, eventHandler) {
    let added = true;
    this.setState((prevState) => {
      const { eventListeners, vaildEventTypes } = prevState;
      if (!vaildEventTypes.includes(eventType)) {
        added = false;
      }
      const index = eventListeners.findIndex( listener => {
        if ((listener.eventHandler === eventHandler) && (listener.eventType == eventType)) {
          return true;
        }
        return false;
      });
      if ( index != -1 ) {
        added = false;
      }
      if (added) {
        eventListeners.push({
          eventType: eventType,
          eventHandler: eventHandler
        });
      }
      return {eventListeners: eventListeners};
    });
    return added;
  }

  removeEventListener(eventType, eventHandler) {
    let deleted = true;
    this.setState((prevState, props) => {
      const { eventListeners, vaildEventTypes } = prevState;
      if (!vaildEventTypes.includes(eventType)) {
        deleted = false;
      }
      const index = eventListeners.findIndex(listener => listener.eventHandler === eventHandler);
      if (index != -1) {
        deleted = false;
      }
      if (deleted) {
        eventListeners.splice(index, 1);
      }
      return {eventListeners: eventListeners};
    });
    return deleted;
  }

  render() {
    const { addEventListener, removeEventListener, handleClick, handleScroll, handleResize } = this;
		const context = {
			addEventListener: addEventListener,
			removeEventListener: removeEventListener,
    }
    
    return (
      <EventListenerContext.Provider value={context}>
        <div className="eventlistener-provider" ref={this.eventRef} onClick={handleClick} onScroll={handleScroll} >
            {this.props.children}
        </div>
      </EventListenerContext.Provider>
    );
  }
}

export default EventListenerWrapper;


export const eventListenerProvider = (Component) => {
  class EventListenerProvider extends React.Component<null, null> {
  render () {
      return (
        <EventListenerWrapper>
          <Component {...this.props} />  
        </EventListenerWrapper>
      );
    }
  }

  return EventListenerProvider;
}

export const eventListenerConsumer = (Component) => {
  class EventListenerConsumer extends React.Component<null, null> {
    render () {
      return (
        <EventListenerContext.Consumer>
          {(context) => {

            return (
              <Component {...this.props} {...context} />
            )
          }}
        </EventListenerContext.Consumer>
      );
    }
  }

  return EventListenerConsumer;
}