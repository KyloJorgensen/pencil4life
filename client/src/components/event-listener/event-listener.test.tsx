'use strict';

import * as React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import { EventListenerWrapper, EventListenerContext, IEventHandler, eventListenerProvider, eventListenerConsumer } from './event-listener';

let wrapper: ShallowWrapper | ReactWrapper;

class Elem extends React.Component<null, null> {
	render() {
		return (
			<p></p>
		)
	}
}

describe('Render EventListenerContext', () => {
	it('matches the snapshot', () => {
		wrapper = mount(
			<EventListenerContext.Consumer>
				{context => {
					return (
						<p></p>
					);
				}}
			</EventListenerContext.Consumer>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('should have a p', () => {
		wrapper = mount(
			<EventListenerContext.Consumer>
				{context => {
					return (
						<p></p>
					);
				}}
			</EventListenerContext.Consumer>
		);
		const child = wrapper.find('p');
		expect(child.length).toEqual(1);
	});

	it('should have a p that contains "child"', () => {
		wrapper = mount(
			<EventListenerContext.Consumer>
				{context => {
					return (
						<p>child</p>
					);
				}}
			</EventListenerContext.Consumer>
		);
		const child = wrapper.find('p').contains('child');
		expect(child).toEqual(true);
	});
});

describe('Render Event component', () => {
	it('matches the snapshot', () => {
		wrapper = shallow(<EventListenerWrapper />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should have a p', () => {
		wrapper = shallow(
			<EventListenerWrapper>
				<p>child</p>
			</EventListenerWrapper>
		);
		const child = wrapper.find('p');
		expect(child.length).toEqual(1);
	});

	it('should have a p that contains "child"', () => {
		wrapper = shallow(
			<EventListenerWrapper>
				<p>child</p>
			</EventListenerWrapper>
		);
		const child = wrapper.find('p').contains('child');
		expect(child).toEqual(true);
	});
});

describe('Render eventListenerProvider', () => {
	it('matches the snapshot', () => {
		const EP = eventListenerProvider(Elem);
		wrapper = shallow(<EP />);
		expect(wrapper).toMatchSnapshot();
	});
});

describe('Render eventListenerConsumer', () => {
	it('matches the snapshot', () => {
		const EC = eventListenerConsumer(Elem);
		wrapper = shallow(<EC />);
		expect(wrapper).toMatchSnapshot();
	});
});

describe('interaction', () => {
	let uniqueNumber = 0;
	let getUniqueNumber = ():number => {
		uniqueNumber++;
		return uniqueNumber;
	};

	interface IUnmountState {
		mount: boolean;
	}
	
	class Unmount extends React.Component<null, IUnmountState> {
		constructor(props) {
			super(props);
			this.state = {
				mount: true,
			}
			this.handleClick = this.handleClick.bind(this);
			this.handleScroll = this.handleScroll.bind(this);
		}

		handleClick(event) {
			this.setState((prevState) => {
				return {mount: !prevState.mount};
			});
		}

		handleScroll(event) {
			this.setState((prevState) => {
				return {mount: !prevState.mount};
			});
		}

		render() {
			return (
				<div>
					<button className="unmount" onClick={this.handleClick} onScroll={this.handleScroll} ></button>
					<div>{this.state.mount ? this.props.children : ''}</div>
				</div>
			);
		}
	}


	interface IInsideProps {
		addEventListener: EventListenerWrapper["addEventListener"];
		removeEventListener: EventListenerWrapper["removeEventListener"];
		clickHandler: IEventHandler;
		scrollHandler: IEventHandler;
	}	
	interface IInsideState {
		uniqueNumber: number;
	}
	class Inside extends React.Component<IInsideProps, IInsideState> {
		constructor(props) {
			super(props)
			this.state = {
				uniqueNumber: getUniqueNumber()
			}
		}

		componentDidMount() {
			this.props.addEventListener('click', this.props.clickHandler);
			this.props.addEventListener('scroll', this.props.scrollHandler);
		}

		componentWillUnmount() {
			this.props.removeEventListener('click', this.props.clickHandler);
			this.props.removeEventListener('scroll', this.props.scrollHandler);
		}

		render() {
			return (
				<p></p>
			);
		}
	}
	beforeEach(() => {
		uniqueNumber = 0;
	});

	it('should add onClick event listener', () => {
		const mockHandler = jest.fn(() => {});
 		wrapper = mount(
			<EventListenerWrapper>
				<EventListenerContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							clickHandler: mockHandler,
							scrollHandler: () => {},
						}

						return (
							<Inside {...props} />
						);
					}}
				</EventListenerContext.Consumer>
			</EventListenerWrapper>
		);

		

		wrapper.find('p').simulate('click');
		expect(mockHandler.mock.calls.length).toEqual(1);
	});

	it('should remove onClick event listener', () => {
		const mockHandler = jest.fn(() => {});
		let props;
 		wrapper = mount(
			<EventListenerWrapper>
				<EventListenerContext.Consumer>
					{context => {
						props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							clickHandler: () => {
								mockHandler();
							},
							scrollHandler: () => {},
						}

						return (
							<Unmount>
								<Inside {...props} />
							</Unmount>
						);
					}}
				</EventListenerContext.Consumer>
			</EventListenerWrapper>
		);
		wrapper.find('button.unmount').simulate('scroll');
		wrapper.find('div').first().simulate('click');
		expect(mockHandler.mock.calls.length).toEqual(0);
	});

	it('should not trigger click on the outside', () => {
		const mockHandler = jest.fn(() => {});
 		wrapper = mount(
			<main>
				<EventListenerWrapper>
					<EventListenerContext.Consumer>
						{context => {
							const props = {
								addEventListener: context.addEventListener,
								removeEventListener: context.removeEventListener,
								clickHandler: mockHandler,
								scrollHandler: () => {},
							}

							return (
								<Inside {...props} />
							);
						}}
					</EventListenerContext.Consumer>
				</EventListenerWrapper>
			</main>
		);

		wrapper.find('main').simulate('click');
		expect(mockHandler.mock.calls.length).toEqual(0);
	});

	it('should trigger all "click" events on click', () => {
		const mockHandler1 = jest.fn(() => {});
		const mockHandler2 = jest.fn(() => {});
		const mockHandler3 = jest.fn(() => {});
 		wrapper = mount(
			<EventListenerWrapper>
				<EventListenerContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							clickHandler: mockHandler1,
							scrollHandler: () => {},
						}

						return (
							<Inside {...props} />
						);
					}}
				</EventListenerContext.Consumer>
				<EventListenerContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							clickHandler: mockHandler2,
							scrollHandler: () => {},
						}

						return (
							<Inside {...props} />
						);
					}}
				</EventListenerContext.Consumer>
				<EventListenerContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							clickHandler: mockHandler3,
							scrollHandler: () => {},
						}

						return (
							<Inside {...props} />
						);
					}}
				</EventListenerContext.Consumer>
			</EventListenerWrapper>
		);

		wrapper.find('p').first().simulate('click');
		expect(mockHandler1.mock.calls.length).toEqual(1);
		expect(mockHandler2.mock.calls.length).toEqual(1);
		expect(mockHandler3.mock.calls.length).toEqual(1);
	});

	it('should add On Scroll event listener', () => {
		const mockHandler = jest.fn(() => {});
 		wrapper = mount(
			<EventListenerWrapper>
				<EventListenerContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							clickHandler: () => {},
							scrollHandler: mockHandler,
						}

						return (
							<Inside {...props} />
						);
					}}
				</EventListenerContext.Consumer>
			</EventListenerWrapper>
		);

		wrapper.find('p').simulate('scroll');
		expect(mockHandler.mock.calls.length).toEqual(1);
	});

	it('should be same when using just the wrappper or EventProvider', () => {
		const EP = eventListenerProvider(Elem);
		wrapper = mount(
			<EP />
		);

		const conatins = wrapper.contains(
			<EventListenerWrapper>
				<Elem />
			</EventListenerWrapper>
		);

		expect(conatins).toEqual(true);
	});
});