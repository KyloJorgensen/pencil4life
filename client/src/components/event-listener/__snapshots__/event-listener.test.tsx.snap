// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Event component matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EventListenerWrapper />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <styled.div>
        <div
          onClick={[Function]}
          onScroll={[Function]}
        />
      </styled.div>,
      "value": Object {
        "addEventListener": [Function],
        "removeEventListener": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div
          onClick={[Function]}
          onScroll={[Function]}
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "onClick": [Function],
          "onScroll": [Function],
        },
        "ref": Object {
          "current": null,
        },
        "rendered": null,
        "type": "div",
      },
      "type": [Function],
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
        "_currentValue2": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
        "_defaultValue": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <styled.div>
          <div
            onClick={[Function]}
            onScroll={[Function]}
          />
        </styled.div>,
        "value": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <div
            onClick={[Function]}
            onScroll={[Function]}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "onClick": [Function],
            "onScroll": [Function],
          },
          "ref": Object {
            "current": null,
          },
          "rendered": null,
          "type": "div",
        },
        "type": [Function],
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "addEventListener": [Function],
            "removeEventListener": [Function],
          },
          "_currentValue2": Object {
            "addEventListener": [Function],
            "removeEventListener": [Function],
          },
          "_defaultValue": Object {
            "addEventListener": [Function],
            "removeEventListener": [Function],
          },
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Render EventListenerContext matches the snapshot 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Context.Consumer>
    [Function]
  </Context.Consumer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": <p />,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": Array [
      undefined,
    ],
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": <p />,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": Array [
        undefined,
      ],
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Render eventListenerConsumer matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EventListenerConsumer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "Consumer": [Circular],
      "Provider": Object {
        "$$typeof": Symbol(react.provider),
        "_context": [Circular],
      },
      "_calculateChangedBits": null,
      "_changedBits": 0,
      "_changedBits2": 0,
      "_currentRenderer": null,
      "_currentRenderer2": null,
      "_currentValue": Object {
        "addEventListener": [Function],
        "removeEventListener": [Function],
      },
      "_currentValue2": Object {
        "addEventListener": [Function],
        "removeEventListener": [Function],
      },
      "_defaultValue": Object {
        "addEventListener": [Function],
        "removeEventListener": [Function],
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
        "_currentValue2": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
        "_defaultValue": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Render eventListenerProvider matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <EventListenerProvider />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <Elem />,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {},
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <Elem />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
